dll/shellext/cabview/folder.cpp

```diff
@@ -0,0 +1,859 @@
+                if (SUCCEEDED(pSF->GetAttributesOf(1, &child, &attrib)))
+                    enable = (attrib & query) == wanted;
+                pSF->Release();
+            }
+            if (enable)
+            {
+                // We don't trust .zip folders, check the file-system to make sure
+                UINT attrib = SUCCEEDED(GetFsPathFromIDList(pidl, buf)) ? GetFileAttributes(buf) : 0;
+                enable = (attrib & FILE_ATTRIBUTE_DIRECTORY) && attrib != INVALID_FILE_ATTRIBUTES;
+            }
+            PostMessageW(hwnd, BFFM_ENABLEOK, 0, enable);
+            break;
+        }
+    }
+    return 0;
+}
+
+struct EXTRACTFILESDATA
+{
+    CCabFolder *pLifetimeCF;
+    HWND hWndOwner;
+    CIDA *pCIDA;
+    STGMEDIUM cidamedium;
+    IDataObject *pDO;
+    IStream *pMarshalDO;
+    IProgressDialog *pPD;
+    UINT cabfiles, completed;
+    WCHAR path[MAX_PATH], cab[MAX_PATH];
+};
+
+static HWND GetUiOwner(const EXTRACTFILESDATA &data)
+{
+    HWND hWnd;
+    if (SUCCEEDED(IUnknown_GetWindow(data.pPD, &hWnd)) && IsWindowVisible(hWnd))
+        return hWnd;
+    return IsWindowVisible(data.hWndOwner) ? data.hWndOwner : NULL;
+}
+
+static HRESULT CALLBACK ExtractFilesCallback(EXTRACTCALLBACKMSG msg, const EXTRACTCALLBACKDATA &ecd, LPVOID cookie)
+{
+    EXTRACTFILESDATA &data = *(EXTRACTFILESDATA*)cookie;
+    switch ((UINT)msg)
+    {
+        case ECM_BEGIN:
+        {
+            data.cabfiles = (UINT)(SIZE_T)ecd.pfdin->hf;
+            return S_OK;
+        }
+
+        case ECM_FILE:
+        {
+            if (data.pPD && data.pPD->HasUserCancelled())
+                return HRESULT_FROM_WIN32(ERROR_CANCELLED);
+            HRESULT hr = data.pCIDA ? S_FALSE : S_OK; // Filtering or all items?
+            if (hr != S_OK)
+            {
+                CABITEM *needle = CreateItem(ecd.pfdin->psz1, ecd.pfdin->attribs);
+                if (!needle)
+                    return E_OUTOFMEMORY;
+                for (UINT i = 0; i < data.pCIDA->cidl && hr == S_FALSE; ++i)
+                {
+                    C_ASSERT(FLATFOLDER);
+                    LPCITEMIDLIST pidlChild = ILFindLastID(HIDA_GetPIDLItem(data.pCIDA, i));
+                    CABITEM *haystack = CABITEM::Validate(pidlChild);
+                    if (!haystack && FAILED_UNEXPECTEDLY(hr = E_FAIL))
+                        break;
+                    if (!lstrcmpiW(needle->Path, haystack->Path))
+                    {
+                        if (data.pPD)
+                            data.pPD->SetLine(1, needle->Path, TRUE, NULL);
+                        hr = S_OK; // Found it in the list of files to extract
+                    }
+                }
+                SHFree(needle);
+            }
+            if (data.pPD)
+                data.pPD->SetProgress(data.completed++, data.cabfiles);
+            return hr;
+        }
+
+        case ECM_PREPAREPATH:
+        {
+            UINT flags = SHPPFW_DIRCREATE | SHPPFW_IGNOREFILENAME;
```

Comment:
Not sure this requires an explicit variable, these could be directly specified in the function call.

Reply:
It can but then you would complain the line is too long. I think this looks better than splitting the function call over two lines.